strictyaml/yaml.go

```diff
@@ -0,0 +1,52 @@
+// Package strictyaml provides a strict YAML unmarshaller based on `go-yaml/yaml`
+package strictyaml
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+
+	"gopkg.in/yaml.v3"
+)
+
+// Unmarshal takes a byte array and an interface passed by reference. The
+// d.Decode will read the next YAML-encoded value from its input and store it in
+// the value pointed to by yamlObj. Any config keys from the incoming YAML
+// document which do not correspond to expected keys in the config struct will
+// result in errors.
+//
+// TODO(https://github.com/go-yaml/yaml/issues/639): Replace this function with
+// yaml.Unmarshal once a more ergonomic way to set unmarshal options is added
+// upstream.
+func Unmarshal(b []byte, yamlObj interface{}) error {
+	r := bytes.NewReader(b)
+
+	d := yaml.NewDecoder(r)
+	d.KnownFields(true)
+
+	// d.Decode will mutate yamlObj
+	err := d.Decode(yamlObj)
+
+	// As bytes are read, the length of the byte buffer should decrease. If it
+	// doesn't, there's a problem.
+	if r.Len() != 0 {
+		return fmt.Errorf("yaml object of size %d bytes had %d bytes of unexpected unconsumed trailers", r.Size(), r.Len())
+	}
+
+	// Show potential error other than EOF which we'll handle later.
+	if err != nil && !errors.Is(err, io.EOF) {
+		return err
+	}
+
+	// When Decode() attempts to parse an empty buffer (config file), an io.EOF is ultimately returned.
+	// 1) Parsing until the end of file returns a nil:
+	//    https://github.com/go-yaml/yaml/blob/f6f7691b1fdeb513f56608cd2c32c51f8194bf51/decode.go#L160-L162
+	// 2) The Decode() checks for that nil and returns the io.EOF
+	//    https://github.com/go-yaml/yaml/blob/f6f7691b1fdeb513f56608cd2c32c51f8194bf51/yaml.go#L123-L126
+	if errors.Is(err, io.EOF) {
+		return fmt.Errorf("yaml object is empty: %s", err)
+	}
```

Comment:
There are some opportunities to condense this code a bit and ensure that we don't accidentally return with the wrong error.

1. We should check the buffer only after we know that the decoding has succeeded. If our caller had passed empty bytes our buffer would also be empty and we would give them the wrong error here.
1. We can do our `err != nil` check just once. Then we can check for `io.EOF` and return our special error indicating that we got nil bytes.
1. Then we can do our buffer check since we've ruled out empty bytes being passed in.

```suggestion
	err := d.Decode(yamlObj)
	if err != nil {
		// io.EOF is returned when the YAML document is empty.
		if errors.Is(err, io.EOF) {
			return fmt.Errorf("unmarshalling YAML, bytes cannot be nil: %w", err)
		}
		return fmt.Errorf("unmarshalling YAML: %w", err)
	}

	// If bytes remain in the reader after decoding, the YAML document contained
	// unexpected bytes.
	if r.Len() != 0 {
		return fmt.Errorf("unmarshalling YAML, found %d unexpected bytes", r.Len())
	}

```

Reply:
@aarongable while working through the above comment I also wrote a couple tests to attempt to trigger the condition you mention in https://github.com/letsencrypt/boulder/pull/6652#discussion_r1107712658

I honestly couldn't find a test case which triggered this. I believe the YAML parser is consuming any parsable document instead of leaving non-YAML lines behind.

Reply:
Bump @aarongable.

Reply:
I agree that Samantha's suggestion here is a more elegant way to accomplish the same goal. I also agree that being super precise about this isn't necessary: we don't parse YAML that often; when we do we're generally only doing it once per process (i.e. not once per request; this isn't performance-sensitive); and it's not even clear that we can run into this error condition anyway. I'm fine with the change landing as-is, or with samantha's suggestion incorporated.

Reply:
I'm going to merge this code as-is. If one day we choose to expand the scope of `//strictyaml` to enforce a certain YAML style then the `bytes.NewReader` length check might come in handy. Currently we'll continue relying on `go-yaml/yaml` swallowing all the bytes it wants.